File: python/mirage/core/wandb/read.py
import json
from collections.abc import AsyncIterator

from mirage.accessor.wandb import WandbAccessor
from mirage.cache.index import NULL_INDEX, IndexCacheStore
from mirage.core.render.json import json_bytes, jsonl_bytes
from mirage.core.wandb.metadata import run_metadata
from mirage.core.wandb.pathing import LEAVES, parts, run_vars
from mirage.core.wandb.queries import RUN, RUN_CONFIG, RUN_SUMMARY
from mirage.core.wandb.stat import stat
from mirage.types import FileType, PathSpec
from mirage.utils.errors import eisdir, enoent


async def read_stream(
        accessor: WandbAccessor,
        path: PathSpec,
        index: IndexCacheStore = NULL_INDEX) -> AsyncIterator[bytes]:
    ps = parts(accessor, path)
    if len(ps) > 4 and ps[3] == "files":
        file = await accessor.client.file(run_vars(ps), "/".join(ps[4:]))
        if file is not None:
            async for chunk in accessor.client.download(
                    file.get("directUrl") or file["url"]):
                yield chunk
            return
    if len(ps) != 4 or ps[3] not in LEAVES:
        info = await stat(accessor, path, index)
        if info.type == FileType.DIRECTORY:
            raise eisdir(path)
        raise enoent(path)
    variables = run_vars(ps)
    if ps[3] == "history.jsonl":
        async for row in accessor.client.history(variables):
            yield jsonl_bytes([row])
        return
    query = {
        "run.json": RUN,
        "config.json": RUN_CONFIG,
        "summary.json": RUN_SUMMARY
    }[ps[3]]
    run = await accessor.client.run(variables, query)
    if ps[3] == "run.json":
        yield json_bytes(run_metadata(run, variables))
    elif ps[3] == "config.json":
        raw = run.get("config") or "{}"
        config = json.loads(raw) if isinstance(raw, str) else raw
        yield json_bytes({k: v["value"] for k, v in config.items()})
    else:
        raw = run.get("summaryMetrics") or "{}"
        yield json_bytes(json.loads(raw) if isinstance(raw, str) else raw)


async def read(accessor: WandbAccessor,
               path: PathSpec,
               index: IndexCacheStore = NULL_INDEX) -> bytes:
    return b"".join(
        [chunk async for chunk in read_stream(accessor, path, index)])

File: typescript/packages/core/src/core/wandb/read.ts
import type { WandbAccessor } from '../../accessor/wandb.ts'
import type { IndexCacheStore } from '../../cache/index/store.ts'
import { FileType, type PathSpec } from '../../types.ts'
import { eisdir, enoent } from '../../utils/errors.ts'
import { jsonBytes, jsonlBytes } from '../render/json.ts'
import { LEAVES, parts, runVars } from './pathing.ts'
import { RUN, RUN_CONFIG, RUN_SUMMARY } from './queries.ts'
import { runMetadata } from './metadata.ts'
import { stat } from './stat.ts'
export async function* readStream(
  accessor: WandbAccessor,
  path: PathSpec,
  index?: IndexCacheStore,
): AsyncGenerator<Uint8Array> {
  const ps = parts(accessor, path)
  if (ps.length > 4 && ps[3] === 'files') {
    const file = await accessor.client.file(runVars(ps), ps.slice(4).join('/'))
    if (file) {
      yield* accessor.client.download(
        file.directUrl === undefined || file.directUrl === null || file.directUrl === ''
          ? file.url
          : file.directUrl,
      )
      return
    }
  }
  if (ps.length !== 4 || !LEAVES.includes(ps[3] ?? '')) {
    const info = await stat(accessor, path, index)
    if (info.type === FileType.DIRECTORY) throw eisdir(path)
    throw enoent(path)
  }
  const variables = runVars(ps)
  if (ps[3] === 'history.jsonl') {
    for await (const row of accessor.client.history(variables)) yield jsonlBytes([row])
    return
  }
  const query = ps[3] === 'run.json' ? RUN : ps[3] === 'config.json' ? RUN_CONFIG : RUN_SUMMARY
  const run = await accessor.client.run(variables, query)
  if (ps[3] === 'run.json') yield jsonBytes(runMetadata(run, variables))
  else if (ps[3] === 'config.json') {
    const raw = run.config === '' ? '{}' : (run.config ?? '{}')
    const config =
      typeof raw === 'string' ? (JSON.parse(raw) as Record<string, { value: unknown }>) : raw
    yield jsonBytes(Object.fromEntries(Object.entries(config).map(([k, v]) => [k, v.value])))
  } else {
    const raw = run.summaryMetrics === '' ? '{}' : (run.summaryMetrics ?? '{}')
    yield jsonBytes(typeof raw === 'string' ? JSON.parse(raw) : raw)
  }
}
export async function read(
  accessor: WandbAccessor,
  path: PathSpec,
  index?: IndexCacheStore,
): Promise<Uint8Array> {
  const chunks: Uint8Array[] = []
  let size = 0
  for await (const chunk of readStream(accessor, path, index)) {
    chunks.push(chunk)
    size += chunk.length
  }
  const result = new Uint8Array(size)
  let offset = 0
  for (const chunk of chunks) {
    result.set(chunk, offset)
    offset += chunk.length
  }
  return result
}
